core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -320,33 +340,73 @@ object ZRef extends Serializable {
         type S = self.S
         def getEither(s: S): Either[ED, D] =
           self.getEither(s).fold(e => Left(eb(e)), bc)
-        def setEither(c: C): Either[EC, S] =
-          ca(c).flatMap(a => self.setEither(a).fold(e => Left(ea(e)), Right(_)))
+        def setEither(c: C)(s: S): Either[EC, S] =
+          ca(c).flatMap(a => self.setEither(a)(s).fold(e => Left(ea(e)), Right(_)))
         val value: Atomic[S] =
           self.value
       }
 
-    final def fold0[EC, ED, C, D](
+    final def foldS[EC >: EB, ED, C, D](
       ea: EA => EC,
-      eb: EB => Either[ED, D],
-      ca: C => Either[EC, A],
+      eb: EB => ED,
+      ca: C => B => Either[EC, A],
       bc: B => Either[ED, D]
     ): ZRef[EC, ED, C, D] =
       new Derived[EC, ED, C, D] {
         type S = self.S
         def getEither(s: S): Either[ED, D] =
-          self.getEither(s).fold(eb, bc)
-        def setEither(c: C): Either[EC, S] =
-          ca(c).flatMap(a => self.setEither(a).fold(e => Left(ea(e)), Right(_)))
+          self.getEither(s).fold(e => Left(eb(e)), bc)
+        def setEither(c: C)(s: S): Either[EC, S] =
+          self.getEither(s).flatMap(ca(c)).flatMap(a => self.setEither(a)(s).fold(e => Left(ea(e)), Right(_)))
         val value: Atomic[S] =
           self.value
       }
 
     final def set(a: A): IO[EA, Unit] =
-      setEither(a).fold(ZIO.fail(_), value.set)
+      value.modify { s =>
+        setEither(a)(s) match {
+          case Left(e)  => (Left(e), s)
+          case Right(s) => (Right(()), s)
+        }
+      }.absolve
 
     final def setAsync(a: A): IO[EA, Unit] =
-      setEither(a).fold(ZIO.fail(_), value.setAsync)
+      value.modify { s =>
+        setEither(a)(s) match {
+          case Left(e)  => (Left(e), s)
+          case Right(s) => (Right(()), s)
+        }
+      }.absolve
+  }
+
+  implicit class LensSnytax[E, A, B](private val self: ZRef[E, E, A, B]) extends AnyVal {
+    def accessField[C, D](lens: ZLens[B, A, D, C]): ZRef[E, E, C, D] =
+      self.foldS(identity, identity, c => b => Right(lens._2(c)(b)), b => Right(lens._1(b)))
+  }
+
+  implicit class OptionalSyntax[A, B](private val self: ZRef[Unit, Unit, A, B]) extends AnyVal {
+    def accessElement[C, D](optional: ZOptional[B, A, D, C]): ZRef[Unit, Unit, C, D] =
+      self.foldS(
+        identity,
+        identity,
+        c => b => Right(optional._2(c)(b)),
+        b => optional._1(b).fold[Either[Unit, D]](Left(()))(Right(_))
+      )
+  }
+
+  implicit class PrismSyntax[E, A, B](private val self: ZRef[E, Unit, A, B]) extends AnyVal {
+    def accessCase[C, D](prism: ZPrism[B, A, D, C]): ZRef[E, Unit, C, D] =
+      self.fold(identity, identity, c => Right(prism._2(c)), b => prism._1(b).fold[Either[Unit, D]](Left(()))(Right(_)))
+  }
+
+  implicit class TraversalSyntax[A, B](private val self: ZRef[Unit, Unit, A, B]) extends AnyVal {
```

Comment:
This one in particular feels like it could benefit from a more polymorphic encoding of the optics.